### ISSUE TITLE
can't construct gradient of gradient

### ISSUE THREAD

**Issue by whdc** [tensorflow/tensorflow#3356]
In general, I've had no trouble constructing the gradient of a gradient in Tensorflow, but I hit on a weird edge case involving complex numbers where I couldn't.  Below, I show a case where the double gradient works, and one that is minimally different where it doesn't.

Here's where double gradient works:
```python
x = tf.Variable(tf.zeros([10]))
xx = x * x
g = tf.gradients(tf.reduce_sum(xx), x)
h = tf.gradients(tf.reduce_sum(g), x)
```
As expected, evaluating `h` produces:
```
>>> sess.run(h, feed_dict={x:10*[1.]})
[array([ 2.,  2.,  2.,  2.,  2.,  2.,  2.,  2.,  2.,  2.], dtype=float32)]
```

Here's where double gradient doesn't work:
```python
x = tf.Variable(tf.zeros([10]))
xx = tf.real(tf.complex(x*x, tf.zeros([10])))
g = tf.gradients(tf.reduce_sum(xx), x)
h = tf.gradients(tf.reduce_sum(g), x)
```
Running the last line gives:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/tensorflow/python/ops/gradients.py", line 494, in gradients
    in_grad.set_shape(t_in.get_shape())
  File "/usr/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 404, in set_shape
    self._shape = self._shape.merge_with(shape)
  File "/usr/local/lib/python2.7/site-packages/tensorflow/python/framework/tensor_shape.py", line 570, in merge_with
    (self, other))
ValueError: Shapes (10,) and () are not compatible
```

I'm running version 0.9.0, CPU only, in OS X.

**Comment by concretevitamin** [comment]
@girving Can you take a look?

**Comment by girving** [comment]
In the complex case `h = tf.gradients(tf.reduce_sum(g), x)` is trying to take the gradient of 2 reals w.r.t. 10 reals, which TensorFlow doesn't support.  We only support gradients of single real scalars.  I believe `h = tf.gradients(tf.real(tf.reduce_sum(g)), x)` should work fine.

**Comment by whdc** [comment]
@girving, `xx` is a tensor of reals, so `g` is a gradient of a real w.r.t. reals.  Supporting this is the fact that the expression `tf.real(tf.reduce_sum(g))` produces the following error:
```
>>> tf.real(tf.reduce_sum(g))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/tensorflow/python/ops/math_ops.py", line 398, in real
    return gen_math_ops.real(input, Tout=input.dtype.real_dtype, name=name)
  File "/usr/local/lib/python2.7/site-packages/tensorflow/python/ops/gen_math_ops.py", line 1516, in real
    result = _op_def_lib.apply_op("Real", input=input, Tout=Tout, name=name)
  File "/usr/local/lib/python2.7/site-packages/tensorflow/python/ops/op_def_library.py", line 530, in apply_op
    _Attr(op_def, input_arg.type_attr))
  File "/usr/local/lib/python2.7/site-packages/tensorflow/python/ops/op_def_library.py", line 61, in _SatisfiesTypeConstraint
    ", ".join(dtypes.as_dtype(x).name for x in allowed_list)))
TypeError: DataType float32 for attr 'T' not in list of allowed values: complex64, complex128
```

**Comment by girving** [comment]
Whoops, you're right.  Apologies for misreading.

**Comment by concretevitamin** [comment]
Assigning to @mrry to take a look, although he's currently out.

**Comment by mrry** [comment]
I think this was fixed by 946f515, which doesn't appear to be included in r0.9. It doesn't reproduce for me at HEAD, but it does with the 0.9 binary. Can you try upgrading to a nightly (or to the imminent r0.10rc0) and see if this is still a problem?

**Comment by mrry** [comment]
I'm going to assume this is fixed now, and close the issue. Feel free to reopen if the problem persists!
